// fm-app-settings/src/utils/settings.js
const NAMESPACE = 'fm-app-settings'

const ID = () => `_${Math.random().toString(36).substr(2, 9)}`

export class Session {
  constructor(namespace) {
    this.namespace = namespace
  }

  defaultNamespace() {
    // session based unique namespace
    let namespace = window.localStorage.getItem(`${ NAMESPACE }.namespace`)
    if (!namespace) {
      namespace = ID()
    }
    return namespace
  }

  get namespace() {
    return this._namespace
  }

  set namespace(namespace) {
    if (!namespace) {
      namespace = this.defaultNamespace()
    }
    this._namespace = namespace
    window.localStorage.setItem(`${ NAMESPACE }.namespace`, namespace)
  }

  getMemoryKey(key) {
    return `${ NAMESPACE }.${ this.namespace }.${key}`
  }

  save(key, value) {
    window.localStorage.setItem(this.getMemoryKey(key), JSON.stringify(value))
  }

  memory(key) {
    let value = window.localStorage.getItem(this.getMemoryKey(key))
    if (value) {
      try {
        return JSON.parse(value)
      } catch(e) {}
    }
  }

  forget(key) {
    window.localStorage.removeItem(this.getMemoryKey(key))
  }
}
