// fm-app-settings/rollup.config.js
import vue from 'rollup-plugin-vue'
import alias from '@rollup/plugin-alias'
import resolve from '@rollup/plugin-node-resolve'
import peerDepsExternal from 'rollup-plugin-peer-deps-external'
import commonjs from '@rollup/plugin-commonjs'
import css from 'rollup-plugin-css-only'

const path = require('path')
const customResolver = resolve({
  extensions: ['.js', '.vue']
})

const plugins = [
  peerDepsExternal(),
  alias({
    entries: [
      {
        find: '~',
        replacement: path.resolve(__dirname, 'src')
      }
    ]
  }),
  resolve({
    extensions: ['.js', '.vue']
  }),
  commonjs(),
  vue({
    css: true,
    style: {
      preprocessOptions: {
        scss: {
          includePaths: ['node_modules']
        }
      }
    }
  }),
  css({ 
    output: 'dist/fm-app-settings.css' 
  })
]

export default [
  {
    input: 'src/index.js',
    output: {
      format: 'esm',
      file: 'dist/fm-app-settings.esm.js'
    },
    plugins
  },

  // SSR build.
  {
    input: 'src/index.js',
    output: {
      format: 'cjs',
      file: 'dist/fm-app-settings.ssr.js'
    },
    plugins
  },

  {
    input: 'src/index.js',
    output: {
      format: 'umd',
      name: 'FmAppSettings',
      file: 'dist/fm-app-settings.umd.js'
    },
    plugins
  }
]
